File: api/v1/schemas/screening.py
from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class CreateScreeningRunRequest(BaseModel):
    trade_date: Optional[date] = Field(None, description="交易日，默认当天")
    stock_codes: Optional[List[str]] = Field(None, description="可选的自定义股票池")
    mode: Optional[Literal["balanced", "aggressive", "quality"]] = Field(None, description="筛选模式预设，默认读取配置")
    candidate_limit: Optional[int] = Field(None, ge=1, le=200, description="规则候选上限，默认读取配置")
    ai_top_k: Optional[int] = Field(None, ge=0, le=50, description="AI 二筛上限，默认读取配置")
    strategies: Optional[List[str]] = Field(None, description="指定使用的筛选策略名称列表，默认使用全部可用策略")
    rerun_failed: bool = Field(False, description="是否在同日同模式失败任务存在时原地补跑")
    resume_from: Optional[Literal["ingesting", "factorizing"]] = Field(None, description="失败任务补跑起始阶段")
    market: Literal["cn"] = Field("cn", description="市场标识，MVP 仅支持 A 股")


class ScreeningNotifyRequest(BaseModel):
    limit: int = Field(10, ge=1, le=50, description="推送候选数量上限")
    with_ai_only: bool = Field(False, description="是否仅推送已进入 AI 二筛的候选")
    force: bool = Field(False, description="是否强制补发（仅对 failed/skipped 有效，v1 不允许重发 sent）")


# ── 五层决策上下文快照 ─────────────────────────────────────────────────────

class MarketEnvironmentSnapshot(BaseModel):
    """L1 大盘环境快照。"""
    market_regime: Optional[str] = None
    risk_level: Optional[str] = None
    index_name: str = "上证指数"
    index_price: Optional[float] = None
    index_ma100: Optional[float] = None
    is_safe: Optional[bool] = None
    message: Optional[str] = None


class SectorHeatSnapshotItem(BaseModel):
    """L2 单个板块热度快照。"""
    board_name: str
    board_type: str = "concept"
    sector_hot_score: float = 0.0
    sector_status: str = "cold"
    sector_stage: str = "ferment"
    canonical_theme: Optional[str] = None
    stock_count: int = 0
    up_count: int = 0
    limit_up_count: int = 0


class DecisionContextSnapshot(BaseModel):
    """L1/L2 决策上下文，附在 ScreeningRunResponse 中返回。"""
    market_environment: Optional[MarketEnvironmentSnapshot] = None
    sector_heat_results: List[SectorHeatSnapshotItem] = Field(default_factory=list)
    hot_theme_count: int = 0
    warm_theme_count: int = 0


class ThemePipelineThemeItem(BaseModel):
    name: str
    normalized_name: Optional[str] = None
    raw_name: Optional[str] = None
    raw_names: List[str] = Field(default_factory=list)
    source: Optional[str] = None
    priority_source: Optional[str] = None
    matched_sources: List[str] = Field(default_factory=list)
    heat_score: Optional[float] = None
    confidence: Optional[float] = None
    source_board: Optional[str] = None
    sector_status: Optional[str] = None
    sector_stage: Optional[str] = None
    stock_count: Optional[int] = None
    up_count: Optional[int] = None
    limit_up_count: Optional[int] = None
    catalyst_summary: Optional[str] = None
    keyword_count: Optional[int] = None
    keywords: List[str] = Field(default_factory=list)
    normalization_status: Optional[str] = None
    normalization_confidence: Optional[float] = None
    normalization_match_reasons: List[str] = Field(default_factory=list)
    normalization_matched_boards: List[str] = Field(default_factory=list)


class LocalThemePipelineSnapshot(BaseModel):
    source: str
    trade_date: Optional[str] = None
    market: Optional[str] = None
    hot_theme_count: int = 0
    warm_theme_count: int = 0
    selected_theme_names: List[str] = Field(default_factory=list)
    themes: List[ThemePipelineThemeItem] = Field(default_factory=list)


class ExternalThemePipelineSnapshot(BaseModel):
    source: str
    trade_date: Optional[str] = None
    market: Optional[str] = None
    accepted_theme_count: int = 0
    hot_theme_count: int = 0
    focus_theme_count: int = 0
    top_theme_names: List[str] = Field(default_factory=list)
    themes: List[ThemePipelineThemeItem] = Field(default_factory=list)


class FusedThemePipelineSnapshot(BaseModel):
    trade_date: Optional[str] = None
    market: Optional[str] = None
    active_sources: List[str] = Field(default_factory=list)
    selected_theme_names: List[str] = Field(default_factory=list)
    merged_theme_count: int = 0
    merged_themes: List[ThemePipelineThemeItem] = Field(default_factory=list)


class ScreeningRunResponse(BaseModel):
    run_id: str
    mode: Optional[str] = None
    status: str
    trade_date: Optional[str] = None
    market: Optional[str] = None
    universe_size: int = 0
    candidate_count: int = 0
    ai_top_k: int = 0
    error_summary: Optional[str] = None
    failed_symbols: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    sync_failure_ratio: float = 0.0
    config_snapshot: Dict[str, Any] = Field(default_factory=dict)
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    # Notification lifecycle fields
    trigger_type: Optional[str] = "manual"
    notification_status: Optional[str] = None
    notification_attempts: int = 0
    notification_sent_at: Optional[str] = None
    notification_error: Optional[str] = None
    # 五层决策上下文
    strategy_names: Optional[List[str]] = None
    decision_context: Optional[DecisionContextSnapshot] = None
    local_theme_pipeline: Optional[LocalThemePipelineSnapshot] = None
    external_theme_pipeline: Optional[ExternalThemePipelineSnapshot] = None
    fused_theme_pipeline: Optional[FusedThemePipelineSnapshot] = None


class ScreeningRunListResponse(BaseModel):
    total: int
    items: List[ScreeningRunResponse]


class ScreeningCandidateItem(BaseModel):
    code: str
    name: Optional[str] = None
    rank: int
    rule_score: float
    selected_for_ai: bool
    matched_strategies: List[str] = Field(default_factory=list)
    strategy_scores: Dict[str, float] = Field(default_factory=dict)
    rule_hits: List[str] = Field(default_factory=list)
    factor_snapshot: Dict[str, Any] = Field(default_factory=dict)
    ai_query_id: Optional[str] = None
    ai_summary: Optional[str] = None
    ai_operation_advice: Optional[str] = None
    has_ai_analysis: Optional[bool] = None
    news_count: Optional[int] = None
    news_summary: Optional[str] = None
    recommendation_source: Optional[str] = None
    recommendation_reason: Optional[str] = None
    final_score: Optional[float] = None
    final_rank: Optional[int] = None
    # -- 五层系统新增字段 (Phase 1) --
    trade_stage: Optional[str] = None
    setup_type: Optional[str] = None
    strategy_family: Optional[str] = None
    entry_maturity: Optional[str] = None
    risk_level: Optional[str] = None
    market_regime: Optional[str] = None
    market_message: Optional[str] = None
    environment_ok: Optional[bool] = None
    index_price: Optional[float] = None
    index_ma100: Optional[float] = None
    theme_position: Optional[str] = None
    theme_tag: Optional[str] = None
    theme_score: Optional[float] = None
    sector_strength: Optional[float] = None
    theme_duration: Optional[str] = None
    trade_theme_stage: Optional[str] = None
    leader_stocks: List[str] = Field(default_factory=list)
    front_stocks: List[str] = Field(default_factory=list)
    candidate_pool_level: Optional[str] = None
    leader_score: Optional[float] = None
    relative_strength_market: Optional[float] = None
    relative_strength_sector: Optional[float] = None
    setup_freshness: Optional[float] = None
    setup_hit_reasons: List[str] = Field(default_factory=list)
    trade_plan: Optional[Dict[str, Any]] = None
    ai_review: Optional[Dict[str, Any]] = None
    # -- AI Review Protocol (Phase 3B-1) --
    ai_trade_stage: Optional[str] = None
    ai_reasoning: Optional[str] = None
    ai_confidence: Optional[float] = None
    ai_environment_ok: Optional[bool] = None
    ai_theme_alignment: Optional[bool] = None
    ai_entry_quality: Optional[str] = None
    stage_conflict: Optional[bool] = None


class ScreeningAnalysisHistoryRef(BaseModel):
    id: Optional[int] = None
    query_id: str
    stock_code: str
    stock_name: Optional[str] = None
    report_type: Optional[str] = None
    operation_advice: Optional[str] = None
    trend_prediction: Optional[str] = None
    sentiment_score: Optional[int] = None
    analysis_summary: Optional[str] = None
    created_at: Optional[str] = None


class ScreeningCandidateDetailResponse(ScreeningCandidateItem):
    analysis_history: Optional[ScreeningAnalysisHistoryRef] = None


class ScreeningCandidateListResponse(BaseModel):
    total: int
    items: List[ScreeningCandidateItem]


class ScreeningStrategyInfo(BaseModel):
    name: str
    display_name: str
    description: str
    category: str
    has_screening_rules: bool
    system_role: Optional[str] = None
    strategy_family: Optional[str] = None
    applicable_market: List[str] = Field(default_factory=list)
    applicable_theme: List[str] = Field(default_factory=list)
    setup_type: Optional[str] = None


class ScreeningStrategyListResponse(BaseModel):
    strategies: List[ScreeningStrategyInfo]

File: apps/dsa-web/src/types/screening.ts
/**
 * 筛选模块类型定义
 * 与后端 API Schema 对齐
 */

// ============ 策略 ============

export interface ScreeningStrategy {
  name: string;
  displayName: string;
  description: string;
  category: 'trend' | 'pattern' | 'reversal' | 'framework';
  hasScreeningRules: boolean;
}

export interface ScreeningStrategyListResponse {
  strategies: ScreeningStrategy[];
}

// ============ 筛选运行 ============

export type ScreeningMode = 'balanced' | 'aggressive' | 'quality';

export type ScreeningRunStatus =
  | 'pending'
  | 'resolving_universe'
  | 'ingesting'
  | 'factorizing'
  | 'screening'
  | 'ai_enriching'
  | 'completed'
  | 'completed_with_ai_degraded'
  | 'failed';

export interface CreateScreeningRunRequest {
  tradeDate?: string;
  stockCodes?: string[];
  mode?: ScreeningMode;
  candidateLimit?: number;
  aiTopK?: number;
  strategies?: string[];
  rerunFailed?: boolean;
  resumeFrom?: 'ingesting' | 'factorizing';
  market?: string;
}

export interface ScreeningRun {
  runId: string;
  mode?: string;
  status: ScreeningRunStatus;
  tradeDate?: string;
  market?: string;
  universeSize: number;
  candidateCount: number;
  aiTopK: number;
  errorSummary?: string;
  failedSymbols: string[];
  warnings: string[];
  syncFailureRatio: number;
  configSnapshot: Record<string, unknown>;
  startedAt?: string;
  completedAt?: string;
  triggerType?: string;
  notificationStatus?: string;
  notificationAttempts: number;
  notificationSentAt?: string;
  notificationError?: string;
  strategyNames?: string[];
  decisionContext?: DecisionContext;
}

export interface ScreeningRunListResponse {
  total: number;
  items: ScreeningRun[];
}

// ============ 五层决策系统类型 ============

export type MarketRegime = 'aggressive' | 'balanced' | 'defensive' | 'stand_aside';
export type ThemePosition = 'main_theme' | 'secondary_theme' | 'follower_theme' | 'fading_theme' | 'non_theme';
export type TradeStage = 'stand_aside' | 'watch' | 'focus' | 'probe_entry' | 'add_on_strength' | 'reject';
export type EntryMaturity = 'low' | 'medium' | 'high';
export type CandidatePoolLevel = 'leader_pool' | 'focus_list' | 'watchlist';
export type SetupType =
  | 'bottom_divergence_breakout'
  | 'low123_breakout'
  | 'trend_breakout'
  | 'trend_pullback'
  | 'gap_breakout'
  | 'limitup_structure'
  | 'none';

export interface TradePlan {
  initialPosition?: string;
  addRule?: string;
  stopLossRule?: string;
  takeProfitPlan?: string;
  invalidationRule?: string;
  riskLevel?: string;
  holdingExpectation?: string;
}

export interface MarketEnvironmentSnapshot {
  marketRegime?: MarketRegime;
  riskLevel?: string;
  indexName?: string;
  indexPrice?: number;
  indexMa100?: number;
  isSafe?: boolean;
  message?: string;
}

export interface SectorHeatSnapshot {
  boardName: string;
  boardType: string;
  sectorHotScore: number;
  sectorStatus: string;
  sectorStage: string;
  canonicalTheme?: string;
  stockCount: number;
  upCount: number;
  limitUpCount: number;
}

export interface DecisionContext {
  marketEnvironment?: MarketEnvironmentSnapshot;
  sectorHeatResults: SectorHeatSnapshot[];
  hotThemeCount: number;
  warmThemeCount: number;
}

// ============ 候选 ============

export interface ScreeningCandidate {
  code: string;
  name?: string;
  rank: number;
  ruleScore: number;
  selectedForAi: boolean;
  ruleHits: string[];
  factorSnapshot: ScreeningFactorSnapshot;
  aiQueryId?: string;
  aiSummary?: string;
  aiOperationAdvice?: string;
  hasAiAnalysis?: boolean;
  newsCount?: number;
  newsSummary?: string;
  recommendationSource?: string;
  recommendationReason?: string;
  finalScore?: number;
  finalRank?: number;
  matchedStrategies?: string[];
  // 五层决策系统字段
  tradeStage?: TradeStage;
  setupType?: SetupType;
  entryMaturity?: EntryMaturity;
  riskLevel?: string;
  marketRegime?: MarketRegime;
  themePosition?: ThemePosition;
  candidatePoolLevel?: CandidatePoolLevel;
  tradePlan?: TradePlan;
  // AI Review Protocol
  aiTradeStage?: string;
  aiReasoning?: string;
  aiConfidence?: number;
}

export interface ScreeningCandidateListResponse {
  total: number;
  items: ScreeningCandidate[];
}

export interface ScreeningCandidateDetail extends ScreeningCandidate {
  analysisHistory?: {
    id?: number;
    queryId: string;
    stockCode: string;
    stockName?: string;
    reportType?: string;
    operationAdvice?: string;
    trendPrediction?: string;
    sentimentScore?: number;
    analysisSummary?: string;
    createdAt?: string;
  };
}

export interface HotThemeNewsItem {
  title: string;
  source?: string;
  summary?: string;
  url?: string;
  published_at?: string;
  heat_score?: number;
}

export interface ScreeningPhaseResults {
  phase1_market_and_theme?: boolean;
  phase2_leader_screen?: boolean;
  phase3_core_signal?: boolean;
  phase4_entry_readiness?: boolean;
  phase5_risk_controls?: boolean;
}

export interface ScreeningPhaseExplanation {
  phase_key: keyof ScreeningPhaseResults;
  label: string;
  hit: boolean;
  summary: string;
}

export interface ScreeningRiskParams {
  stop_loss?: number;
  position_size?: string;
  take_profit_ratio?: number;
}

export interface TechnicalPatternMetric {
  readonly label: string;
  readonly value: string;
}

export interface TechnicalPattern {
  readonly id: string;
  readonly name: string;
  readonly signalStrength?: number;
  readonly metrics: readonly TechnicalPatternMetric[];
  readonly hitReasons: readonly string[];
}

export interface ScreeningFactorSnapshot extends Record<string, unknown> {
  close?: number;
  volume_ratio?: number;
  pct_chg?: number;
  is_hot_theme_stock?: boolean;
  primary_theme?: string;
  theme_heat_score?: number;
  theme_match_score?: number;
  leader_score?: number;
  extreme_strength_score?: number;
  entry_reason?: string;
  core_signal?: string;
  theme_catalyst_summary?: string;
  theme_catalyst_news?: HotThemeNewsItem[];
  phase_results?: ScreeningPhaseResults;
  phase_explanations?: ScreeningPhaseExplanation[];
  risk_params?: ScreeningRiskParams;
  extreme_strength_reasons?: string[];
  bottomDivergenceHitReasons?: string[];
  bottom_divergence_hit_reasons?: string[];
  rule_hits_display?: string[];
  bonus_signals?: string[];

  // Bottom divergence pattern fields
  bottom_divergence_double_breakout?: boolean;
  bottom_divergence_state?: string;
  bottom_divergence_pattern_code?: string;
  bottom_divergence_pattern_label?: string;
  bottom_divergence_signal_strength?: number;
  bottom_divergence_entry_price?: number;
  bottom_divergence_stop_loss?: number;
  bottom_divergence_horizontal_breakout?: boolean;
  bottom_divergence_trendline_breakout?: boolean;
  bottom_divergence_sync_breakout?: boolean;

  // Pattern 123 fields
  pattern_123_low_trendline?: boolean;
  pattern_123_state?: string;
  pattern_123_entry_price?: number;
  pattern_123_stop_loss?: number;
  pattern_123_signal_strength?: number;

  // MA100+Low123 combined fields
  ma100_low123_confirmed?: boolean;
  ma100_low123_pattern_strength?: number;
  ma100_low123_ma_score?: number;
  ma100_low123_hit_reasons?: string[];

  // MA100+60min combined fields
  ma100_60min_confirmed?: boolean;
  ma100_60min_freshness_score?: number;
  ma100_60min_ma_score?: number;
  ma100_60min_hit_reasons?: string[];

  // Simple boolean signal flags
  above_ma100?: boolean;
  gap_breakaway?: boolean;
  is_limit_up?: boolean;
}

// ============ 通知 ============

export interface ScreeningNotifyRequest {
  limit?: number;
  withAiOnly?: boolean;
  force?: boolean;
}

// ============ 辅助 ============

export const SCREENING_STAGES: ScreeningRunStatus[] = [
  'resolving_universe',
  'ingesting',
  'factorizing',
  'screening',
  'ai_enriching',
  'completed',
];

export const STAGE_LABELS: Record<string, string> = {
  pending: '等待中',
  resolving_universe: '解析股票池',
  ingesting: '同步数据',
  factorizing: '构建因子',
  screening: '规则筛选',
  ai_enriching: 'AI 分析',
  completed: '已完成',
  completed_with_ai_degraded: '已完成(AI降级)',
  failed: '失败',
};

export const CATEGORY_LABELS: Record<string, string> = {
  trend: '趋势',
  pattern: '形态',
  reversal: '反转',
  framework: '框架',
};

export const CATEGORY_COLORS: Record<string, string> = {
  trend: 'bg-blue-500/20 text-blue-400 border-blue-500/30',
  pattern: 'bg-purple-500/20 text-purple-400 border-purple-500/30',
  reversal: 'bg-orange-500/20 text-orange-400 border-orange-500/30',
  framework: 'bg-emerald-500/20 text-emerald-400 border-emerald-500/30',
};

/**
 * The 4 target strategies for the consolidated screening page.
 * Strategies not yet implemented are shown as disabled (hasScreeningRules=false).
 */
export const TARGET_STRATEGIES: ScreeningStrategy[] = [
  {
    name: 'bottom_divergence_double_breakout',
    displayName: '底背离双突破',
    description: '基于DIF/DEA六形态底背离 + 下降趋势线/水平阻力线双突破确认',
    category: 'reversal',
    hasScreeningRules: true,
  },
  {
    name: 'ma100_low123_combined',
    displayName: 'MA100+低位123结构',
    description: '站上MA100均线 + 低位123底部反转形态联合确认',
    category: 'reversal',
    hasScreeningRules: true,
  },
  {
    name: 'ma100_60min_combined',
    displayName: 'MA100+60分钟线',
    description: '日线站上MA100 + 60分钟级别入场信号联合确认',
    category: 'trend',
    hasScreeningRules: true,
  },
  {
    name: 'extreme_momentum_combined',
    displayName: '极端强势组合',
    description: '涨停突破/缺口突破等极端强势形态联合策略',
    category: 'trend',
    hasScreeningRules: false,
  },
];

export function getStageIndex(status: ScreeningRunStatus): number {
  const idx = SCREENING_STAGES.indexOf(status);
  return idx >= 0 ? idx : 0;
}

export function isTerminalStatus(status: ScreeningRunStatus): boolean {
  return ['completed', 'completed_with_ai_degraded', 'failed'].includes(status);
}

// ============ 五层决策标签映射 ============

export const TRADE_STAGE_LABELS: Record<string, string> = {
  probe_entry: '试探入场',
  add_on_strength: '加仓确认',
  focus: '重点关注',
  watch: '持续观察',
  stand_aside: '暂时回避',
  reject: '不参与',
};

export const TRADE_STAGE_COLORS: Record<string, string> = {
  probe_entry: 'bg-emerald-500/20 text-emerald-400 border-emerald-500/30',
  add_on_strength: 'bg-green-500/20 text-green-400 border-green-500/30',
  focus: 'bg-blue-500/20 text-blue-400 border-blue-500/30',
  watch: 'bg-gray-500/20 text-gray-400 border-gray-500/30',
  stand_aside: 'bg-orange-500/20 text-orange-400 border-orange-500/30',
  reject: 'bg-red-500/20 text-red-400 border-red-500/30',
};

export const THEME_POSITION_LABELS: Record<string, string> = {
  main_theme: '主力题材',
  secondary_theme: '次要题材',
  follower_theme: '跟风题材',
  fading_theme: '退潮题材',
  non_theme: '非题材',
};

export const THEME_POSITION_COLORS: Record<string, string> = {
  main_theme: 'bg-purple-500/20 text-purple-400 border-purple-500/30',
  secondary_theme: 'bg-blue-500/20 text-blue-400 border-blue-500/30',
  follower_theme: 'bg-cyan-500/20 text-cyan-400 border-cyan-500/30',
  fading_theme: 'bg-orange-500/20 text-orange-400 border-orange-500/30',
  non_theme: 'bg-gray-500/20 text-gray-400 border-gray-500/30',
};

export const MARKET_REGIME_LABELS: Record<string, string> = {
  aggressive: '积极进攻',
  balanced: '均衡配置',
  defensive: '防御为主',
  stand_aside: '空仓观望',
};

export const MARKET_REGIME_COLORS: Record<string, string> = {
  aggressive: 'bg-emerald-500/20 text-emerald-400 border-emerald-500/30',
  balanced: 'bg-blue-500/20 text-blue-400 border-blue-500/30',
  defensive: 'bg-orange-500/20 text-orange-400 border-orange-500/30',
  stand_aside: 'bg-red-500/20 text-red-400 border-red-500/30',
};

export const SETUP_TYPE_LABELS: Record<string, string> = {
  bottom_divergence_breakout: '底背离突破',
  low123_breakout: '低位123突破',
  trend_breakout: '趋势突破',
  trend_pullback: '趋势回调',
  gap_breakout: '跳空突破',
  limitup_structure: '涨停结构',
  none: '无',
};

export const POOL_LEVEL_LABELS: Record<string, string> = {
  leader_pool: '龙头池',
  focus_list: '关注池',
  watchlist: '观察池',
};

export const ENTRY_MATURITY_LABELS: Record<string, string> = {
  high: '成熟',
  medium: '发展中',
  low: '早期',
};
